exonum/src/events/noise/wrappers/sodium_wrapper/wrapper.rs

```diff
@@ -123,55 +128,60 @@ impl NoiseWrapper {
 
-        LittleEndian::write_u32(&mut msg_len_buf, len as u32);
-        let encoded_message = &encrypted_message[0..len];
-        msg_len_buf.extend_from_slice(encoded_message);
-        buf.extend_from_slice(&msg_len_buf);
+            encrypted_message[start..end].copy_from_slice(&written[..len]);
+        }
 
-        debug_assert_eq!(len, encoded_message.len());
-        Ok(None)
+        buf.extend_from_slice(&encrypted_message);
+        Ok(())
     }
 
-    fn read(&mut self, input: &[u8], len: usize) -> Result<Vec<u8>, NoiseError> {
-        let mut buf = vec![0_u8; len];
-        self.session.read_message(input, &mut buf)?;
-        Ok(buf)
+    fn read(&mut self, input: &[u8], buf: &mut [u8]) -> Result<usize, NoiseError> {
+        let len = self.session.read_message(input, buf)?;
+        Ok(len)
     }
 
-    fn write(&mut self, msg: &[u8]) -> Result<Vec<u8>, NoiseError> {
-        let mut buf = vec![0_u8; MAX_MESSAGE_LENGTH];
-        let len = self.session.write_message(msg, &mut buf)?;
-        buf.truncate(len);
-        Ok(buf)
+    fn write(&mut self, msg: &[u8], buf: &mut [u8]) -> Result<usize, NoiseError> {
+        let len = self.session.write_message(msg, buf)?;
+        Ok(len)
     }
 
-    // Each message consists of the payload and 16 bytes(`TAG_LENGTH`)
-    // of AEAD authentication data. Therefore to calculate an actual message
-    // length we need to subtract `TAG_LENGTH` multiplied by messages count
-    // from `data.len()`.
-    fn decrypted_msg_len(&self, raw_message_len: usize) -> usize {
-        raw_message_len - TAG_LENGTH * (raw_message_len / MAX_MESSAGE_LENGTH)
+    fn noise_builder<'a>() -> Builder<'a> {
+        Builder::with_resolver(PARAMS.parse().unwrap(), Box::new(SodiumResolver::new()))
     }
+}
 
-    // In case of encryption we need to add `TAG_LENGTH` multiplied by messages count to
-    // calculate actual message length.
-    fn encrypted_msg_len(&self, raw_message_len: usize) -> usize {
-        raw_message_len + TAG_LENGTH * ((raw_message_len / MAX_MESSAGE_LENGTH) + 1)
-    }
+// Each message consists of the payload and 16 bytes(`TAG_LENGTH`)
+// of AEAD authentication data. Therefore to calculate an actual message
+// length we need to subtract `TAG_LENGTH` multiplied by messages count
+// from `data.len()`.
+fn decrypted_msg_len(raw_message_len: usize) -> usize {
+    raw_message_len - TAG_LENGTH * (div_ceil(raw_message_len, MAX_MESSAGE_LENGTH))
+}
 
-    fn noise_builder<'a>() -> Builder<'a> {
-        Builder::with_resolver(PARAMS.parse().unwrap(), Box::new(SodiumResolver::new()))
+// In case of encryption we need to add `TAG_LENGTH` multiplied by messages count to
+// calculate actual message length.
+fn encrypted_msg_len(raw_message_len: usize) -> usize {
+    raw_message_len
+        + TAG_LENGTH * (raw_message_len.div_floor(&(MAX_MESSAGE_LENGTH - TAG_LENGTH)) + 1)
+}
+
+fn div_ceil(lhs: usize, rhs: usize) -> usize {
+    match lhs.div_rem(&rhs) {
```

Comment:
Nit: I wonder how the use of `div_rem` here compares to using built-in `%` and `/` operations. If I understand correctly, it's the only use of the `num` crate in the main code.

Reply:
I think we can remove it.